crates/edr_provider/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+mod config;
+mod data;
+mod error;
+mod filter;
+mod logger;
+mod requests;
+/// Test utilities
+#[cfg(test)]
+pub mod test_utils;
+
+use data::ProviderData;
+use requests::{
+    eth::{self, handle_evm_mine_request},
+    hardhat,
+};
+use tokio::sync::Mutex;
+
+use self::requests::{EthRequest, Request};
+pub use self::{
+    config::{AccountConfig, ProviderConfig},
+    error::ProviderError,
+    requests::ProviderRequest,
+};
+
+/// A JSON-RPC provider for Ethereum.
+pub struct Provider {
+    data: Mutex<ProviderData>,
+}
+
+// Add a layer in front that handles this
+// let RpcRequest {
+//     version,
+//     method: request,
+//     id,
+// } = request;
+
+// if version != jsonrpc::Version::V2_0 {
+//     return Err(ProviderError::RpcVersion(version));
+// }
+//
+// fn to_response(
+//     id: jsonrpc::Id,
+//     result: Result<serde_json::Value, ProviderError>,
+// ) -> jsonrpc::Response<serde_json::Value> {
+//     let data = match result {
+//         Ok(result) => jsonrpc::ResponseData::Success { result },
+//         Err(error) => jsonrpc::ResponseData::Error {
+//             error: jsonrpc::Error {
+//                 code: -32000,
+//                 message: error.to_string(),
+//                 data: None,
+//             },
+//         },
+//     };
+
+//     jsonrpc::Response {
+//         jsonrpc: jsonrpc::Version::V2_0,
+//         id,
+//         data,
+//     }
+// }
+
+impl Provider {
+    pub async fn handle_request(
+        &self,
+        request: ProviderRequest,
+    ) -> Result<serde_json::Value, ProviderError> {
+        let mut data = self.data.lock().await;
+
+        // TODO: resolve deserialization defaults using data
+        // Will require changes to `ProviderRequest` to receive `json_serde::Value`
+
+        match request {
+            ProviderRequest::Single(request) => handle_single_request(&mut data, request).await,
+            ProviderRequest::Batch(requests) => handle_batch_request(&mut data, requests).await,
+        }
+    }
+}
+
+/// Handles a JSON request for an execution provider.
+pub async fn handle_single_request(
+    data: &mut ProviderData,
+    request: Request,
+) -> Result<serde_json::Value, ProviderError> {
+    match request {
+        Request::Eth(request) => handle_eth_request(data, request).await,
+        Request::Hardhat(request) => handle_hardhat_request(data, request).await,
+    }
+}
+
+/// Handles a batch of JSON requests for an execution provider.
+pub async fn handle_batch_request(
```

Comment:
I'm curious why this and the previous function are public.

Reply:
Good catch. Thanks!